README.md
# Network of evolving trading strategies 

Inspired by Tom Schaul's thesis on evolving a compact, concept-based Sokoban solver. 

Generates strategies of the form (AND (indicator & args) (AND (NOT (indicator & args)) (OR (...) (...)))) and simulates them on EURUSD, 1M candles. Merges branches from existing strategies with likelihood 1/(market exposure)^2

Work in progress. Next few updates will include:
* significantly faster simulation by generating table of returns at each candle at various trailing distances
* connecting to InteractiveBroker API for real-time testing and trading
* Adding neural networks to the agnts' toolbelts

## Installation

Be sure install https://github.com/sbhaaf/clj-ta-lib for the technical indicator library. Other deps in project.clj

## Usage

Not currently in format supporting uberjar, but could be easily. 
Open in nREPL and spin up workers to start testing strategies.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-12-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-12-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hayek/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hayek/compare/0.1.0...0.1.1


test/hayek/core_test.clj
(ns hayek.core-test
  (:require [clojure.test :refer :all]
            [hayek.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/hayek/core.clj
(ns hayek.core
  (:require [com.hypirion.clj-xchart :as c])
  (:use [hayek.ta-lib]
        [hayek.utils]
        [hayek.simulator]
        [hayek.strategies])
  (:gen-class))


;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;  Helper Functions  ;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;

(do
  
  (defn reset-agntlog [agntlog]
    (dosync
     (ref-set agntlog {:max_id -1
                       :agnts []}))
    nil)

  (defn save-agntlog
    [agntlog]
    (spit "resources/agnts/agnt.log" @agntlog))


  (defn load-agntlog []
    (ref (read-string (slurp "resources/agnts/agnt.log"))))


  (defn generate-agnt
    ([params] (generate-agnt params (random-fn-tree params)))
    ([params fn_tree]
     {:fn_tree fn_tree
      :fn (compile-fn-tree fn_tree)
      :wallet 1000
      :atr_candles 14
      :atr_dist (+ 1 (* 3 (rand)))
      :closing_fn 'trailing-stop-ATR
      :bid_fn 'only-long-bids
      :parents nil
      :archive []
      }))


  (defn log-agnt-performace
    [agnt tradelog agntlog]
    (if (not (some #{0} (map count tradelog)))
      (let [agnt_id (ref -1)
            num_trades (map count tradelog)
            risk (map #(abs (mean (filter (partial > 0) (map :gain %)))) tradelog)
            reward (map #(mean (filter (partial < 0) (map :gain %))) tradelog)
            num_won (map #(count (filter (partial < 0) (map :gain %))) tradelog)
            num_lost (map #(count (filter (partial > 0) (map :gain %))) tradelog)
            exposure (map #(cond (= 0 (nth num_won %)) 10
                                 (= 0 (nth num_lost %)) 1
                                 :else
                                 (/ (nth risk %) (nth reward %)
                                    (/ (nth num_won %) (nth num_lost %))))
                          (range (count tradelog)))]
        (if (= 0 (count exposure))
          (do (def tradelog tradelog)
              (def agnt agnt)
              (println exposure)
              -1)
          (do (dosync
               (let [id (inc (agntlog :max_id))]
                 (ref-set agnt_id id)
                 (alter agntlog assoc
                        :max_id id
                        :agnts (conj (@agntlog :agnts)
                                     {:id id
                                      :exposure exposure
                                      :num_trades num_trades}))))
              (save-agntlog agntlog)
              (spit (apply str "resources/agnts/" @agnt_id ".agnt")
                    {:agnt agnt
                     :id @agnt_id
                     :num_trades (map count tradelog)
                     :risk risk
                     :reward reward
                     :num_won num_won
                     :num_lost num_lost
                     :exposure exposure
                     })
              )))
      -1))

  
  (defn tradelog-from-results
    [agnt periods results]
    (loop [periods periods
           tradelog []]
      (let [bids ((eval (agnt :bid_fn)) agnt (first periods))
            trades (loop [trades []
                          results results
                          bids bids]
                     (if (= 0 (count results))
                       trades
                       (if (not= (first bids) [1 0 0])
                         (recur (conj trades (first results)) (rest results) (rest bids))
                         (recur trades (rest results) (rest bids)))))]
        (recur (rest periods) (conj tradelog trades)))))

  
  (defn simulate-all-periods
    [agnt periods agntlog params]
    (if (params :memoize_trades)
      ;; get results from simulating across all periods
      (let [results ((params :memo-trade-fn)
                     periods (agnt :atr_candles) (agnt :atr_dist) (agnt :bid_fn))
            tradelog (tradelog-from-results agnt periods results)]
        (if (params :log_performace)
          (log-agnt-performace agnt tradelog agntlog)
          tradelog))
      (loop [periods periods
             tradelog []]
        (if (= 0 (count periods))
          (if (params :log_performace)
            (log-agnt-performace agnt tradelog agntlog)
            tradelog)
          (if (and (< 0 (count tradelog)) (= 0 (count (last tradelog))))
            -1
            (let [[trades agnt] (simulate-period (first periods) agnt params)]
              (recur (rest periods) (conj tradelog trades))))))))


  (defn generate-and-sim
    [periods agntlog params]
    (let [agnt (generate-agnt params)]
      (simulate-all-periods agnt periods agntlog params)))


  (defn window-exposure
    "Not built for data-stream. Window assumes every agent tested up to current."
    ([n agntlog] (map #(window-exposure n (% :num_trades) (% :exposure))
                      (agntlog :agnts)))
    ([n num_trades exposure]
     (/ (apply + (map #(* (nth num_trades %) (nth exposure %))
                      (range (- (count exposure) n) (count exposure))))
        (apply + (take-last n num_trades)))))


  (defn roullete
    "Takes a vector of magnitudes and returns index n with proportionate likelihood"
    [coll]
    (let [num (count coll)
          val (rand)
          coll (map (partial * (/ 1 (apply + coll))) coll)]
      (loop [sum (first coll)
             coll (rest coll)]
        (if (> sum val)                    
          (dec (- num (count coll)))
          (recur (+ sum (first coll)) (rest coll))))))

  (defn load-agnt
    [id]
    (:agnt (read-string (slurp (apply str "resources/agnts/" id ".agnt")))))



  
  (defn select-random-agnt
    [agntlog params]
    (let [exposure (window-exposure (params :window_size) agntlog)
          id (roullete (power (map (partial / 1) exposure) 2))] ;; maybe ^2
      [id, (load-agnt id)])) 
  

  (defn merge-and-sim
    [periods agntlog params]
    (let [[id1 agnt1] (select-random-agnt agntlog params)
          [id2 agnt2] (select-random-agnt agntlog params)
          fn_tree (merge-fn-trees (agnt1 :fn_tree) (agnt2 :fn_tree) params)
          agnt (assoc (generate-agnt params fn_tree) :parents [id1 id2])]
      (simulate-all-periods agnt periods agntlog params)
      ))
        

  (defn all-trade-results
    [periods atr_candles atr_dist bid_fn]
    (let [agnt {:fn '(fn [period] (repeat (count period) true))
                :atr_candles atr_candles
                :atr_dist atr_dist
                :closing_fn 'trailing-stop-ATR
                :bid_fn bid_fn
                }
          params {:update_agnt false
                  :log_performace false
                  :every_candle true
                  }]
      (simulate-all-periods agnt periods nil params)))

  
  (defn crawl-fn-tree
    ([fn_tree fun] (crawl-fn-tree fn_tree fun [] []))
    ([fn_tree fun loc results]
     (let [node (reduce #(%1 %2) fn_tree loc)]
       (if (contains? node :gate)
         (loop [results results
                i (range ((node :gate) :num-inp))]
           (if (= 0 (count i))
             results
             (recur (apply conj results
                          (crawl-fn-tree fn_tree fun
                                         (conj loc :gate
                                               (keyword (str (first i)))) results))
                    (rest i))))
         [[loc, (fun (node :indicator))]]))))

  
  ;; (defn optimize-fn-tree
    ;; "takes an agent and runs one round of a greedy optimization."
    ;; [periods agnt params]

    ;; (let [params (assoc params :log_performance true)
          ;; fun ]
    
  ;; (fn 

    ;; (def node ((random-fn-tree params) :indicator))
    ;; ))

  
  )
  

;;;;;;;;;;;;;;
;;;  Main  ;;;
;;;;;;;;;;;;;;

;; TODO:
;;   * run/balance multiple modes of agnt generation
;;     X generate new trading strategies
;;       - ratio: 0.5
;;     + merge most successful strategies
;;       - ratio: 0.25
;;     + mutate old strategies
;;       - ratio: 0.10
;;     + optimize old strategies
;;       - ratio: 0.15
;;       - note: if we save multiple, we're increasing likelihood of family exploding
;;         - might be okay
;;   * Run long & short on every candle in every period at several ATR-distances
;;     + will allow each agent to be tested in much shorter time
;;     + pass the expanded results as parameter to closing function
;;       - :every_candle = true/false to determine what index should be returned
;;   * Memoize each function component as well (harder)
;;   * make max_depth for random agnt generation
;;   * stop simulation on agnt if no trades in first period
;;   * 
    



(defn -main
  [& args]
  
  (do
    (def periods (get-periods))
    (def period (first periods))
    (def agntlog (load-agntlog))
    (def params {:update false
                 :log_performace true
                 :leverage 100
                 :gate_likelihood 0.4
                 :node_likelihood 0.7
                 :every_candle false
                 :memoize_trades false
                 :window_size 17
                 :memo-trade-fn (resetable-memoize all-trade-results)
                 })
    )
  (do
    (def new_workers (n-serial-workers 16 'generate-and-sim periods agntlog params))
    (def merge_workers (n-serial-workers 8 'merge-and-sim periods agntlog params)))
  
  (do
    (kill-workers new_workers)
    (kill-workers merge_workers))

  (generate-and-sim periods agntlog params)
  (merge-and-sim periods agntlog params)
  
  )

(do
  (def a agnt)
  (def t tradelog))

(reset-agntlog agntlog)








;; (def chart1
;;          (c/xy-chart {"DI+" [(range 50) (take 50 (first bull))]
;;                       "DI-" [(range 50) (take 50 (first bear))]}))
;; (def chart2
;; (c/xy-chart {"Close" [(range 50)
;;                       (take 50 (drop 13 (first (makeOHLC period "C"))))]}))

;; (c/view chart1 chart2)





src/hayek/utils.clj
(ns hayek.utils
  (:import [java.util Random]))


;;;;;;;;;;;;;;;;;;;;;;
;;;  Random tools  ;;;
;;;;;;;;;;;;;;;;;;;;;;

(defn reduce-by-index
  ([coll] (reduce-by-index '+ 0 coll))
  ([fun coll] (reduce-by-index 'fun 0 coll))
  ([fun val coll]
   (map #(reduce (fn [val item] (fun val (nth item %)))
                 val coll)
        (range 3))))

(defn power [base exp] (apply * (repeat base exp)))

(defn stdev [coll]
  (defn squares [avg coll] (map #(power (- % avg) 2) coll))
  (let [avg (mean coll)
        squares (squares avg coll)
        total (count coll)]
    (Math/sqrt (/ (reduce + squares) (- total 1)))))


(defn round-up [val] (int (Math/ceil val)))


(defn which-max [coll]
  (first (apply max-key second (map-indexed vector coll))))


(defn which-min [coll]
  (first (apply min-key second (map-indexed vector coll))))


(defn update-last [coll & kvs]
  "takes a list of maps followed by key-value pairs"
  (update-in coll [(dec (count coll))]
             #(eval (conj kvs % 'assoc))))


(defn first-index-of "For each item in sub, return index in coll"
  [sub coll]
  (map (fn [s] (first (keep-indexed #(if (= s %2) %1) coll))) sub))


(defn abs [n] (max n (- n)))

(defn mean
  [coll] (if (= 0 (count coll))
                         nil
                         (/ (apply + coll) (count coll))))

(defmacro bench
  "Times the execution of forms, discarding their output and returning
a long in nanoseconds."
  ([& forms]
   `(let [start# (System/nanoTime)]
       [~@forms (/ (- (System/nanoTime) start#) 1000000000.)])))


(defn resetable-memoize
  "Call (<fn> :reset!) to clear the cache"
  [f]
  (let [mem (atom {})]
    (fn [& args]
      (if (= (first args) :reset!)
        (reset! mem {})
        (if-let [e (find @mem args)]
          (val e)
          (let [ret (apply f args)]
            (swap! mem assoc args ret)
            ret))))))

(defn rand-norm 
  ([] (rand-norm 0.0 1.0))
  ([mu sigma]
   (+ mu (* sigma (.nextGaussian (Random.))))))





;;;;;;;;;;;;;;;;;;;;;
;;;  Concurrency  ;;;
;;;;;;;;;;;;;;;;;;;;;

(defn worker
  [fun & args]
  (let [alive (ref true)]
    {:future (future
               (loop []
                 (if @alive
                   (do
                     (apply fun args)
                     (recur))
                   "DEAD")))
    :alive alive}))


(defn kill-workers
  [workers]
  (if (not= (type {}) (type workers))
    (map kill-workers workers)
    (let [alive (workers :alive)]
      (dosync
       (ref-set alive false)))))


(defn n-serial-workers
  "'fun' must be quoted"
  [n fun & args]
  
  (defn update-mean-time
    [mean_time t]
    (dosync (ref-set mean_time
                     [(inc (first @mean_time))
                      (/ (+ (* (first @mean_time)
                               (second @mean_time)) t)
                         (inc (first @mean_time)))])))
  
  (let [mean_time (ref [0 0])]
    (doall (repeatedly n #(worker (fn [fun args mean_time fn_name]
                                    (let [[res t] (bench (apply (eval fun) args))]
                                      (if (not= res -1)
                                        (do
                                          (update-mean-time mean_time t)
                                          (if (= 0 (rem (first @mean_time) 10))
                                            (do
                                              (println fn_name "x" n "workers")
                                              (println "   iterations:" (first @mean_time))
                                              (println "   mean time:"
                                                       (/ (second @mean_time) n)
                                                       "seconds")))))))
                                  fun args mean_time (str fun)
                                  )))
    ))



;;;;;;;;;;;;;;;;;;;;;
;;;  Data import  ;;;
;;;;;;;;;;;;;;;;;;;;;

(defn my-read-string [val]
  (if (= val "") "" (read-string val)))


(defn parse
  "Convert the CSV into rows of columns"
  [string]
  (let [split_string (map #(clojure.string/split % #",") 
                          (clojure.string/split string #"\n"))]
    (apply conj [(first split_string)]
          (map #(apply conj [(first %)] (into [] (map read-string (drop 1 %))))
               (rest split_string)))))


(defn read-data [path]
  (parse (slurp path)))


(defn get-period
  "At the end of each row is a 0/1 that signifies the end of a
   trading period. Return first period and the rest of the data"
  [data]
  (loop [period [(first data)]
         data (rest data)]
    (if (= 1 (last (last period)))
      [period data]
      (recur (conj period (first data)) (rest data)))))


(defn split-periods
  "Split into periods to train on. See get-period."
  [data]
  (loop [periods []
         remaining data]
    (if (> (count remaining) 0)
      (let [[period remaining] (get-period remaining)]

        (recur (conj periods period) remaining))
      periods)))

(defn get-periods
  []
  (loop [periods []
         period []
         bid (rest (read-data "resources/data/EURUSD_1_m_BID.csv"))
         ask (rest (read-data "resources/data/EURUSD_1_m_ASK.csv"))]
    (if (= 0 (count bid))
      periods
      (if (and (= 0 (nth (first bid) 5)) (= 0 (count period))) ;; clear zeros 
        (recur periods period (rest bid) (rest ask))
        (let [row (conj (first bid) (- (nth (first ask) 4) (nth (first bid) 4)))
              period (conj period row)]
          (if (= 0 (apply + (map #(nth % 5) (take 10 (rest bid)))))
            (recur (conj periods period) [] (rest bid) (rest ask))
            (recur periods period (rest bid) (rest ask))))))))





src/hayek/strategies.clj
(ns hayek.strategies
  (:use [hayek.utils]
        [hayek.ta-lib]))


;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;  Helper Functions  ;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;

;; (defn BS->BSH [BS]
;;   (map #(apply conj (vector (abs (apply min %)))
;;                (map (partial - (apply min %)) %)) BS))


;; (defn row-stochastic [coll]
;;   (map (fn [row] (map (fn [idx] (/ idx (apply + row))) row)) coll))


;; (defn equate-atom-lengths [& atoms]
;;   (let [m (apply min (map (comp count deref) atoms))]
;;     (doseq [a atoms]
;;       (swap! a (partial take-last m)))))


;; (defn one-hot
;;   [n coll]
;;   (let [emp (apply vector (repeat n 0))]
;;     (map #(assoc emp % 1) coll)))

(do 
(defn crosses
  "finds the crosses of two vectors"
  [alpha beta]
  (loop [return [0]
         rel (if (> (first alpha) (first beta)) 1 2)
         alpha (rest alpha)
         beta (rest beta)]
    (if (= 0 (count alpha))
      return
      (if (> (first alpha) (first beta))
        (if (= rel 1)
          (recur (conj return 0) 1 (rest alpha) (rest beta))
          (recur (conj return 1) 1 (rest alpha) (rest beta)))
        (if (= rel 2)
          (recur (conj return 0) 2 (rest alpha) (rest beta))
          (recur (conj return 2) 2 (rest alpha) (rest beta)))))))


(defn diff [vec]
  (loop [rem vec
         dif [0]]
    (if (= 1 (count rem))
      dif
      (recur (rest rem) (conj dif (- (second rem) (first rem)))))))


(defn sign [vec]
  (map #(cond (> 0 %) -1 (< 0 %) 1 :else 0) vec))


(defn sign-change [vec]
  (loop [res [0]
         vec vec]
    (if (= 1 (count vec))
      res
      (recur (conj res (cond (and (> (second vec) 0)
                                  (< (first vec) 0)) 1
                             (and (< (second vec) 0)
                                  (> (first vec) 0)) 2
                             :else 0))
             (rest vec)))))


(defn which-greater
  "returns which one is greater"
  [alpha beta]
  (map #(cond (> (nth alpha %) (nth beta %)) 1
              (< (nth alpha %) (nth beta %)) 2
              :else 0) (range (count alpha))))

(defn map_>=
  "returns true where > c"
  [coll c]
  (map #(>= % c) coll))

(defn map_<=
  "returns true where > c"
  [coll c]
  (map #(<= % c) coll))

(defn map_=
  "returns true where > c"
  [coll c]
  (map #(= % c) coll))

(defn map_NOT
  "returns (not %) onto coll"
  [coll]
  (map #(not %) coll))
 
(defn nth_>=
  "returns true where alpha > beta"
  [alpha beta]
  (map #(>= (nth alpha %) (nth beta %)) (range (count alpha))))

(defn nth_<=
  "returns true where alpha > beta"
  [alpha beta]
  (map #(<= (nth alpha %) (nth beta %)) (range (count alpha))))

(defn nth_=
  "returns true where alpha > beta"
  [alpha beta]
  (map #(= (nth alpha %) (nth beta %)) (range (count alpha))))

(defn nth_AND
  "returns true where (and alpha beta). 
   NOTE: responsible for returning with length of shorter indicator"
  [alpha beta]
  (let [len (min (count alpha) (count beta))
        alpha (take-last len alpha)
        beta (take-last len beta)]
    (map #(and (nth alpha %) (nth beta %)) (range (count alpha)))))

(defn nth_OR
  "returns true where (or alpha beta)
   NOTE: responsible for returning with length of shorter indicator"
  [alpha beta]
  (let [len (min (count alpha) (count beta))
        alpha (take-last len alpha)
        beta (take-last len beta)]
    (map #(or (nth alpha %) (nth beta %)) (range (count alpha)))))

(defn binned
  "returns integers corresponding to bins"
  [coll knots]
  (map (fn [obs] (or (first (filter #(> (nth knots %) obs)
                                    (range (count knots))))
                     (count knots)))
       coll))

(defn get-random-gate
  [gates]
  (rand-nth gates))

(defn get-AND-gate [gates] (first gates))
(defn get-OR-gate [gates] (second gates))





;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;  Strategy fragments  ;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;


(def indicators
  [{:name "DI"
    :params [['timeperiod 14]]
    :data ['DI+ 'DI-]
    :methods [{:desc "which-greater"
               :fn '(fn [pos neg] (which-greater pos neg))
               :type :flag
               :range [1 2]}
              {:desc "which-crossed"
               :fn '(fn [pos neg] (crosses pos neg))
               :type :flag
               :range [1 2]}
              ]}
   {:name "ADX"
    :params [['timeperiod 14]]
    :data ['ADX]
    :methods [{:desc "which-bin"
               :fn '(fn [d] (binned d [15 20 25 30]))
               :type :scale
               :range [0 1 2 3 4]}
              ]}
   {:name "MACD"
    :params [['fastperiod 12] ['slowperiod 26] ['signalperiod 9]]
    :data ['MACD]
    :methods [{:desc "which-crossed"
               :fn '(fn [macd signal hist] (crosses signal macd))
               :type :flag
               :range [1 2]
               }
              {:desc "sign of angle"
               :fn '(fn [macd signal hist] (sign (diff hist)))
               :type :flag
               :range [-1 1]
               }
              ]}
   {:name "SAR"
    :params [['acceleration 0] ['maximum 0]] 
    :data ['SAR]
    :methods [{:desc "above/below price"
               :fn '(fn [period sar]
                      (let [open (take-last (count sar) (map (fn [r] (nth r 1)) period))]
                        (sign (map (fn [n] (- (nth open n) (nth sar n)))
                                              (range (count sar))))))
               :period true
               :type :flag
               :range [-1 1]
               }
              {:desc "crossed-price"
               :fn '(fn [period sar]
                      (let [open (take-last (count sar) (map (fn [r] (nth r 1)) period))]
                        (crosses open sar)))
               :period true
               :type :flag
               :range [1 2]
               }
              ]}
   {:name "CMO"
    :params [['timeperiod 14]]
    :data ['CMO]
    :methods [{:desc "below/within/above -50/50"
               :fn '(fn [d] (binned d [-50 50]))
               :type :flag
               :range [0 1 2]
               }
              ]}
   {:name "ADL"
    :params [['fastperiod 9] ['slowperiod 16]]
    :data ['ADL]
    :methods [{:desc "sign"
               :fn 'sign
               :type :flag
               :range [-1 1]
               }
              {:desc "sign-change"
               :fn 'sign-change
               :type :flag
               :range [1 2]
               }
              ]}
   {:name "Aroon"
    :params [['timeperiod 20]]
    :data ['Aroon]
    :methods [{:desc "one above and other below 50"
               :fn '(fn [pos neg upper lower]
                      (map (fn [n]
                             (cond (and (> (nth pos n) upper) (< (nth neg n) lower)) 1
                                   (and (> (nth neg n) upper) (> (nth pos n) lower)) 2
                                   :else 0)) (range (count pos))))
               :params [['upper 50] ['lower 50]]
               :type :flag
               :range [1 2]
               }
              ]}
   {:name "CCI"
    :params [['timeperiod 20]]
    :data ['CCI]
    :methods [{:desc "reversal at extreme"
               :fn '(fn [cci thresh]
                      (loop [res [0]
                             cci cci]
                        (if (= 1 (count cci)) res
                          (recur (conj res (cond (and (> (first cci) thresh)
                                                      (< (second cci) thresh)) 2
                                                 (and (< (first cci) (* -1 thresh))
                                                      (> (second cci) (* -1 thresh))) 1
                                                 :else 0)) (rest cci)))))
               :params [['thresh 100]]
               :type :flag
               :range [1 2]
               }
              ]}
   ;; {:name "RSI"
    ;; :params [['timeperiod 14]]
    ;; :data ['RSI]
    ;; }
   ])

;; ;; Just a reminder
;; (def indicators
;;   X {:fn 'SAR
;;   ?  :role :trailing_stop}
;;    {:fn 'Keltner
;;     :role :band} ; error
;;    {:fn 'ATR
;;     :role :volatility}
;;    {:fn 'ROC
;;     :role :trend_confirmation}
;;    {:fn 'StDev
;;     :role :volatility}
;;    {:fn 'RSI
;;     :role :buying_pressure}

(def gates
  [{:name :AND
    :num-inp 2
    :fn 'nth_AND}
   {:name :OR
    :num-inp 2
    :fn 'nth_OR}
   ;; {:name :NOT
    ;; :num-inp 1
    ;; :fn 'map_NOT}
   ;; {:name :BUT-NOT
    ;; :num-inp 2
    ;; :fn '(and %1 (not %2))}
   ])


(def protocols
  {:flag [{:desc "flag_="
           :fn 'map_=
           :param-fns ['(fn [range] (rand-nth range))]}]
   :bool [{:desc "is-true"
           :fn '%
           :param-fns []}]
   :scale [{:desc "scale_="
            :fn 'map_=
            :param-fns ['(fn [range] (rand-nth range))]}
           {:desc "scale_>="
            :fn 'map_>=
            :param-fns ['(fn [range] (rand-nth (drop-last (drop 1 range))))]}
           {:desc "scale_<="
            :fn 'map_<=
            :param-fns ['(fn [range] (rand-nth (drop-last (drop 1 range))))]}]
   })



;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;  Strategy generator  ;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; TODO:
;;   X When a new strategy is initiated, only combine with AND
;;   * Determine whether NOT is important, or if it can be acheived naturally
;;     + probably can be worked to produce same likelihood with less adverse effects.
;;   X New strategies should be combined in several ways at once
;;     + AND and OR gates on all or fragments of the trees
;;       - This should be done at the top-level. merge fn takes specified gate.
;;   * New strategies (and potentially old ones) should be optimized
;;     + should be done to all active strategies on current window
;;     + need to write a method that moves parameters towards optimal
;;       - Try n manipulations and choose greedily
;;   * Mutually exclusive strategies fragments should be weeded out
;;   * Parameters right now are pulled from normal distribution
;;     + when the default is 0 there needs to be a new procedure
;;   * Handle cases where fast MUST BE < slow
;;   * make random parameter option for METHODS, i.e. CCI and Aroon

;; Considerations:
;;   * To NOT or not to NOT?
;;     + to NOT:
;;       - running :scale on bands or equivalent might utilize NOT-within or NOT-outside
;;       - merging agnts of opposite directions
;;     + not to NOT:
;;       -
;;     + maybe use NOT only when combining opposing strategies



(defn random-fn-tree
  [params]
  
  ;;;  helper functions  ;;;
  (defn use-random-protocol
    [protocols method]
    (let [protocol (rand-nth (protocols (method :type)))]
      (assoc protocol :params (map #(% (method :range))
                                   (eval (protocol :param-fns))))))
  
  (defn get-random-indicator
    [indicators protocols]
    (let [indicator (rand-nth indicators)
          ;; pull integer value for each parameter from normal distribution
          data_args (map #(round-up (max 6 (rand-norm (second %)
                                                      (* (second %) 0.25))))
                         (indicator :params))
          method (rand-nth (indicator :methods))
          method_args (if (contains? method :params)
                        (into [] (map #(round-up (max 6 (rand-norm (second %)
                                                                   (* (second %) 0.25))))
                                      (method :params))))]
      {:name (indicator :name)
       :args data_args
       :data (indicator :data)
       :method (assoc method
                      :protocol (use-random-protocol protocols method)
                      :args method_args)
       }))

  ;;;  Body  ;;;
  (if (> (params :gate_likelihood) (rand))
    (let [gate (get-AND-gate gates)]
      (loop [fn-branch {:gate gate}
             i 0]
        (if (> (inc i) (gate :num-inp))
          fn-branch
          (recur (assoc-in fn-branch [:gate (keyword (str i))]
                        (random-fn-tree params))
                 (inc i)))))
    {:indicator (get-random-indicator indicators protocols)})
  )


(defn compile-fn-tree
  [tree]

  (defn recursive-process
    [tree]
    (if (contains? tree :indicator)
      (let [indicator (tree :indicator)
            method (indicator :method)
            protocol (method :protocol)
            data-fns (map #(conj (indicator :args) 'period %)
                          (indicator :data))
            data-fns (if (and (contains? method :period) (method :period))
                       (concat [['period]] data-fns)
                       data-fns) ;; some methods may utilize raw price
            data-fns (list 'apply 'concat (conj (into [] data-fns) (method :args)))
            ]
        (concat (list (protocol :fn) (list 'apply (method :fn) data-fns))
                (protocol :params)))
      (let [gate (tree :gate)
            args (into (list) (map #(recursive-process (gate (keyword (str %))))
                                   (range (gate :num-inp))))]
        (conj args (gate :fn))
        )))
  
  (into (list (recursive-process tree)) (list '[period] 'fn)))


(defn merge-fn-trees
  "Merge random branches of two trees under gate. Random gate if not specified."
  ([tree1 tree2 params] (merge-fn-trees tree1 tree2 (get-random-gate gates) params))
  ([tree1 tree2 gate params]
   (defn choose-node [tree params]
     (loop [res 0]
       (if (not= 0 res)
         res
         (if (> (params :node_likelihood) (rand))
           tree
           (recur (loop [tree tree]
                    (if (contains? tree :indicator)
                      (if (> (params :node_likelihood) (rand))
                        tree
                        0)
                      (let [nodes (shuffle [((tree :gate) :0)
                                            ((tree :gate) :1)])]
                        (cond (> (params :node_likelihood) (rand))
                              (first nodes)
                              (> (params :node_likelihood) (rand))
                              (second nodes)
                              :else (cond (contains? (first nodes) :gate)
                                          (recur (first nodes))
                                          (contains? (second nodes) :gate)
                                          (recur (second nodes))
                                          :else 0))))))))))
   {:gate (assoc gate :0 (choose-node tree1 params) :1 (choose-node tree2 params))}))

)

;;;;;;;;;;;;;;
;;;  test  ;;;
;;;;;;;;;;;;;;

;; (do
;;    (def periods (get-periods))
;;    (def period (first periods)))


;; (do
;;   (def params {:gate_likelihood 0
;;                :node_likelihood 0.7})
;;   (def tree (random-fn-tree params))
;;   (def tree1 (random-fn-tree params))
;;   (def tree2 (random-fn-tree params))
;;   (compile-fn-tree (merge-fn-trees tree1 tree2 params)))

;; ((eval (compile-fn-tree tree)) period)

;; (repeatedly 1000
;;             #(do
;;                (def tree (random-fn-tree params))
;;                ((eval (compile-fn-tree tree))period)))


